### PR TITLE
fix(prometheus): ensure metrics are lazily loaded

### PR DESCRIPTION
**Pull request by aarnphm**
This PR address the issue inside the container, where if the user service uses `bentoml.metrics`, the service is imported before `PROMETHEUS_MULTIPROC_DIR` is created.

Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

### CODE REVIEW

**Comment by bojiang** [comment]
@aarnphm Why do we need this?

**Comment by aarnphm** [comment]
When users use `bentoml.metrics` inside service.py, this would currently fail, since the multiproc dir is not yet initialized (inside container)

File: src/bentoml/metrics.py [hunk 1]
```diff
@@ -13,8 +20,34 @@ def __dir__() -> list[str]:
...
+        self._kwargs: dict[str, t.Any] = {}
+
+    def __call__(self, *args: t.Any, **kwargs: t.Any) -> t.Any:
+        # This is where we lazy load the proxy object.
+        self._args = args
+        self._kwargs = kwargs
+        return self
+
+    def __getattr__(self, item: t.Any) -> t.Any:
+        if self._proxy is None:
```
Comment (bojiang) [inline src/bentoml/metrics.py:38]:
```suggestion
    def __getattr__(self, item: t.Any) -> t.Any:
        if item in {"_attr", "_proxy", ...}:
           raise AttibuteError()
        if self._proxy is None:
```

File: src/bentoml/metrics.py [hunk 2]
```diff
@@ -13,8 +20,36 @@ def __dir__() -> list[str]:
...
+    @inject
+    def _load_proxy(
+        self,
+        metrics_client: PrometheusClient = Provide[BentoMLContainer.metrics_client],
+    ) -> None:
+        self._proxy = getattr(metrics_client, self._attr)(*self._args, **self._kwargs)
+        self._initialized = True
+
+
 def __getattr__(item: t.Any):
```
Comment (bojiang) [inline src/bentoml/metrics.py:54]: shall we check the typo for users

**Comment by bojiang** [comment]
@aarnphm The test is still failling

**Comment by bojiang** [comment]
It seems the e2e tests are skipped. Will continue the review when it's enabled

**Review by sauyon — APPROVED** [review]

**Pull request merged**